components/action-mgt/org.wso2.carbon.identity.action.execution/src/main/java/org/wso2/carbon/identity/action/execution/ActionExecutorServiceImpl.java

```diff
@@ -0,0 +1,399 @@
+        try {
+            return actionExecutor.get();
+        } catch (InterruptedException | ExecutionException e) {
+            throw new ActionExecutionRuntimeException("Error occurred while executing action: " + action.getId(),
+                    e);
+        }
+    }
+
+    private void logActionRequest(String apiEndpoint, ActionType actionType, String actionId,
+                                  AuthMethods.AuthMethod authenticationMethod, String payload) {
+
+        //todo: Add to diagnostics
+        if (LOG.isDebugEnabled()) {
+            LOG.debug(String.format(
+                    "Calling API: %s for action type: %s action id: %s with authentication: %s payload: %s",
+                    apiEndpoint,
+                    actionType,
+                    actionId,
+                    Optional.ofNullable(authenticationMethod).map(AuthMethods.AuthMethod::getAuthType)
+                            .orElse("NONE"),
+                    payload));
+        }
+    }
+
+    private ActionExecutionStatus processActionResponse(ActionInvocationResponse actionInvocationResponse,
+                                                        ActionType actionType,
+                                                        Map<String, Object> eventContext,
+                                                        ActionExecutionRequest actionRequest,
+                                                        String actionId, String apiEndpoint,
+                                                        AuthMethods.AuthMethod authenticationMethod,
+                                                        ActionExecutionResponseProcessor
+                                                                actionExecutionResponseProcessor)
+            throws ActionExecutionResponseProcessorException {
+
+        if (actionInvocationResponse.isSuccess()) {
+            return processSuccessResponse((ActionInvocationSuccessResponse) actionInvocationResponse.getResponse(),
+                    actionType,
+                    eventContext, actionRequest, actionId, apiEndpoint, authenticationMethod,
+                    actionExecutionResponseProcessor);
+        } else if (actionInvocationResponse.isError() && actionInvocationResponse.getResponse() != null) {
+            return processErrorResponse((ActionInvocationErrorResponse) actionInvocationResponse.getResponse(),
+                    actionType,
+                    eventContext, actionRequest, actionId, apiEndpoint, authenticationMethod,
+                    actionExecutionResponseProcessor);
+        } else {
+            logErrorResponse(actionInvocationResponse, actionType, actionId, apiEndpoint, authenticationMethod);
+        }
+
+        return new ActionExecutionStatus(ActionExecutionStatus.Status.FAILURE, eventContext);
+    }
+
+    private ActionExecutionStatus processSuccessResponse(ActionInvocationSuccessResponse successResponse,
```

Comment:
This method has 8 input parameters. Typically, having more than 4 parameters suggests that the method should be refactored, either through object encapsulation or by splitting it into smaller methods.

In this case, the following call seems to be the reason:

`logSuccessResponse(successResponse, actionType, actionId, apiEndpoint, authenticationMethod);
`

Reply:
yeh will improve to some level. Most probably will have to redefine this contract as we go with different action types. Will improve the rest later.